### ISSUE TITLE
Primitive Data Types: Jdouble

### ISSUE THREAD

**Issue by nsanitate** [apuliasoft/j2se-js#24]
Implement the **double** primitive data type that allow to overcome the differences between Java and JavaScript.
> The double data type is a double-precision 64-bit IEEE 754 floating point. Its range of values is beyond the scope of this discussion, but is specified in the [Floating-Point Types, Formats, and Values](https://docs.oracle.com/javase/specs/jls/se7/html/jls-4.html#jls-4.2.3) section of the Java Language Specification.

## Expected Behavior

* Declaring a variable to refer to a double
* Creating, Initializing, and Accessing a double
  * initialize with `d` or `f`
  * initialize with scientific notation (`e` or `E`)
* Equality and Relational Operators
  * equal to
  * not equal to
  * greater than
  * greater than or equal to
  * less than
  * less than or equal to
* Arithmetic Operators
  * additive operator
  * subtraction operator
  * multiplication operator
  * division operator
  * remainder operator
* Unary Operators
  * unary plus operator
  * unary minus operator
  * increment operator
  * decrement operator
* Exception
  * incompatible types: <type> cannot be converted to double
  
### Declaring a variable to refer to a double
```typescript
let aDouble: Jdouble;
```

### Creating, Initializing, and Accessing a double
```typescript
const aDouble = jdouble('3.1');
console.log("Double: " + aDouble);

// initialize with `d` or `f`
const aDouble = jdouble('3.1d');
console.log("Double with d: " + aDouble);

// initialize with scientific notation (`e` or `E`)
const aDouble = jdouble('3.1e7');
console.log("Double scientific: " + aDouble);
```
Output:
```
Double: 3.1
Double with d: 3.1
Double scientific: 3.1E7
```

### Equality and Relational Operators
```typescript
const value1 = jdouble('1.1');
const value2 = jdouble('2.2');

if(is(value1.eq(value2)))
   console.log("value1 == value2");

if(is(value1.ne(value2)))
   console.log("value1 != value2");

if(is(value1.gt(value2)))
   console.log("value1 > value2");

if(is(value1.ge(value2)))
   console.log("value1 > value2");

if(is(value1.lt(value2)))
   console.log("value1 < value2");

if(is(value1.le(value2)))
   console.log("value1 <= value2");
```
Output:
```
value1 != value2
value1 <  value2
value1 <= value2
```

### Arithmetic Operators
```typescript
const value2 = jdouble('0.2');
const value7 = jdouble('0.7');
let result: Jdouble;

result = value7.add(value2);
console.log("0.7 + 0.2 = " + result);

result = value7.sub(value2);
console.log("0.7 - 0.2 = " + result);

result = value7.mul(value2);
console.log("0.7 * 0.2 = " + result);

result = value7.div(value2);
console.log("0.7 / 0.2 = " + result);

result = value7.mod(value2);
console.log("0.7 % 0.2 = " + result);
```
Output:
```
0.7 + 0.2 = 0.8999999999999999
0.7 - 0.2 = 0.49999999999999994
0.7 * 0.2 = 0.13999999999999999
0.7 / 0.2 = 3.4999999999999996
0.7 % 0.2 = 0.09999999999999992
```

### Unary Operators
```typescript
const value1 = jdouble('-1.1');
let result: Jdouble;

result = value1.plus();
console.log("+(-1.1) = " + result);

value1.inc();
console.log("++(-1,1) = " + value1);

value1.dec();
console.log("--(-0.1) = " + value1);

result = result.minus();
console.log("-(-1.1) = " + result);
```
Output:
```
+(-1.1) = -1.1
++(-1.1) = -0.1
--(-0.1) = -1.1
-(-1.1) = 1.1
```

### Exceptions
```typescript
const aDouble = jdouble('three point one');
```
Output:
```
ERROR: incompatible types: String cannot be converted to double
```

**Comment by dammafra** [comment]
## Questions
1. In Java, char and double can be used interchangeably, ie in Java is possible to do:
    ``` java
    double aDouble = 'c';
    System.out.println(anInt); // prints 99.0
    ```
    Should we replicate? If so, which of the two alternative below?
    ``` javascript
    // first alternative
    const aDouble = jdouble('c');

    // second alternative
    const anotherDouble = jdouble(jchar('c'));
    ```


2. In Java, primitive types do not automatically set a default value for functions local variables, but only for class members; we set always a default value instead. Should we replicate?
3. Exceptions listed in "Expected Behavior" section are compile time errors in Java and not runtime ones. I still have to handle them?
4. In Java, double can be initialized with an expression, ie in Java is possible to do:
    ``` java
    double aDouble = 1.2 + 2.3;
    System.out.println(anInt); // prints 3.5
    ```
    We could easily do this with constructor by number; should we replicate it with constructor by string also? (This could be useful: https://pegjs.org)
    **Note:** In Java, mixed expression initialization like `double aDouble = 1 + 2.0;` are valid; in our case, constructor by number correctly works.
5. In Java, doubles are always printed showing their decimal part or in scientific notation when needed or in according to instantiation mode; in typescript this doesn't happen in the same way. Should we replicate?
    **Possible solution:** Keep track of the instantiation mode  with regex match and boolean flags, to be used in toString method.

**Comment by nsanitate** [comment]
1. Yes, we should implement the constructor
    ``` javascript
    const aDouble: Jdouble = jdouble(jchar('c'));
    ```
2. No, let's use default in every case; e.g. when `let d: Jdouble = jdouble();` the variable `d` is initialized to value `0.0` in case it's a class member or a local variable.
3. Yes, we can't raise this type of exceptions at compile time, so we should use JavaScript error at runtime.
4. Such cases are managed in the following way:
    ``` typescript
    const aDouble: Jdouble = jdouble('1.1').add(jdouble('2.2'));
    ```
5. Don't worry about it, we should study every case before implementing `System.out.print*`. Maybe we could use a special `toString` function that decides how to represent number as string.

**Comment by nsanitate** [comment]
Please, remove constructor by number.

**Comment by dammafra** [comment]
Implemented mixed operations between `Jint` (#23), `Jdouble` and `Jchar` (#26). Maybe the code is too duplicated?

**Comment by nsanitate** [comment]
Please, implement more representative tests.
